Support wildcard patch releases

jafeltra opened the pull request:
This PR adds support for using a wildcard to download the latest patch release of a package. A user can specify a version like `3.0.x`, and FPL will now find the latest patch release of `3.0`.

I think the approach I took to finding and determining the latest patch seems reasonable, but let me know if I missed anything. I also reused the `LatestVersionUnavailableError`, adding in a flag for if it was a patch version, but if anyone thinks it would be cleaner just to separate the errors, I'm happy to. I also realized I didn't include the word `"wildcard"` in the function name, so if anyone thinks that would be helpful for future-us, I'm happy to adjust it.

jafeltra committed:
Support using a wildcard to check for the latest patch release

cmoesel commented on src/errors/LatestVersionUnavailableError.ts:
```diff
@@ -1,7 +1,13 @@
+  constructor(
+    public packageName: string,
+    public customRegistry?: string,
+    public isPatch?: boolean
```
I think `isPatchWildCard` or `isWildCardPatch` would be more descriptive.

jafeltra replied:
[2b59a31](https://github.com/standardhealth/fhir-package-loader/pull/15/commits/2b59a31682443905280984131c2923b69a86b706)

cmoesel commented on src/load.ts:
```diff
@@ -392,6 +395,53 @@ export async function lookUpLatestVersion(
+
+  if (res?.data?.versions) {
+    const versions = Object.keys(res.data.versions);
+    const regex = new RegExp(`^${version.split('.').slice(0, -1).join('\\.')}\\.\\d+$`); // regex for matching major.minor version, excluding any qualifiers (like -draft)
+    const matchingVersions = versions.filter(v => v.match(regex));
+    if (matchingVersions.length === 0) {
+      throw new LatestVersionUnavailableError(packageName, customRegistry, true);
+    }
+    const patchVersionNumbers = matchingVersions.map(v =>
+      parseInt(v.slice(v.lastIndexOf('.') + 1))
+    );
+    const latestPatchIndex = patchVersionNumbers.findIndex(
+      p => p === Math.max(...patchVersionNumbers)
+    );
+    return matchingVersions[latestPatchIndex];
```
This looks like it should work, but it is a little complicated. What do you think about using [semver](https://www.npmjs.com/package/semver)? I think if we used it, all of above would become:
```js
const latest = maxSatisfying(versions, version, false);
if (latest == null) {
  throw new LatestVersionUnavailableError(packageName, customRegistry, true);
}
return latest;
```

I'm also OK w/ keeping as-is, but if we do that, I think we should check for `/\d+\.\d+\.x/` in this function too since your logic assumes that format.

jafeltra replied:
Ohh, no I like that much better. Much simpler. I never thought about using semver.

jafeltra replied:
[25e0135](https://github.com/standardhealth/fhir-package-loader/pull/15/commits/25e01357887179fbd36dac50224ff18fe5d3a865)

jafeltra replied:
Oh wait, I did have a question. I'm not sure I understand the third parameter and why we want to match versions loosely. Is there a scenario I should add a test for to cover it? I just realized I forgot to add that parameter back in but all the tests are currently passing, so maybe I'm missing a test case.

cmoesel replied:
I said that parameter should be `false` -- which means _don't_ match loosely. But on a further look at the docs, the default value is `false` anyway, so the two-arg call is fine (and maybe even better).

jafeltra replied:
Oh. Wow I double confused myself. I did check that the docs said the default was false, but then when I came back here, I looked at the `true` in the error, so I thought "okay so why do I need to explicitly set it to true". Whoops.

cmoesel commented on test/load.test.ts:
```diff
@@ -35,6 +36,39 @@ const TERM_PKG_RESPONSE = {
+      fhirVersion: 'R4',
+      url: 'https://packages.simplifier.net/hl7.terminology.r4/2.0.0'
+    },
+    '2.0.2': {
```
This is a little nit-picky, but... this package's `dist-tags` says the `latest` is `1.2.3-test`, so it's a little weird to have versions `2.x` in the manifest. Maybe they can be `1.1.0`, `1.1.1`, and `1.1.2` instead?

jafeltra replied:
Whoops. I didn't think about that in context. I can swap them around.

jafeltra replied:
[135ecd3](https://github.com/standardhealth/fhir-package-loader/pull/15/commits/135ecd36b1b70c299ba77eafcc19b4726ef91c0d)

cmoesel reviewed (changes requested):
Cool cool cool. I've made a few comments for your consideration. Nothing major, but probably worth considering?

jafeltra committed:
Update error parameter name

jafeltra committed:
Use version numbers that make more sense in context

jafeltra committed:
Use semver to find the latest patch version from the provided versions

jafeltra commented:
> Nothing major, but probably worth considering?

Heh, get it??

I think I've addressed your comments, but let me know if I missed anything or you notice anything else.

cmoesel commented on src/load.ts:
```diff
@@ -392,6 +396,46 @@ export async function lookUpLatestVersion(
+
+  if (res?.data?.versions) {
+    const versions = Object.keys(res.data.versions);
+    const latest = maxSatisfying(versions, version);
```
Should we do one more check to ensure that the version is in the right format (i.e., `/\d+\.\d+\.x/`)? As it is now, if we passed in a version like `2.x`, it would give us the latest minor version. Since the function specifically says latest patch version, and since we export it, I think it makes sense to do the check so we're not false advertising.

jafeltra replied:
Ohh yes, we definitely should.

jafeltra replied:
Okay I looked at this briefly and I feel like this could actually go two ways:
- I could do what you suggested and add the check for the version format at the top of the function to ensure only patch wild cards are passed in.
- Alternatively, I could rename this function to be `lookupLatestWildcardVersion` and not include that check. There is a check in `mergeDependency` that would ensure that implementation only gets wildcard versions for patch wildcards. Part of me likes this approach because now that we're using semver in the function, there's no reason it needs to be specific to patch versions anymore. However, FHIR only allows patch wildcards, so I also understand wanting to restrict the functionality of this function.

I could go either way, so if you have an opinion, let me know.

cmoesel replied:
Yeah, I had thought about that as well, and honestly, I'm also fine with either way. If we went w/ option 2, maybe I'd name it `lookupMaxSatisfyingVersion` or `resolveSemverVersion` -- as it works for any semver (not just wildcards).

jafeltra replied:
@cmoesel  - Today I'm leaning toward the first option, just because the second option makes it seem like we'd potentially support wildcard functionality that we don't have any intention of supporting (because FHIR doesn't support it). Plus, if we did have that function support any semver, maybe we should add tests to ensure that too, and that seems unnecessary for our use case. The first option seems to show the intended functionality more easily.

I'll go with the first option, unless you disagree.

cmoesel replied:
@jafeltra - Works for me!

jafeltra replied:
Done in [41ce0c7](https://github.com/standardhealth/fhir-package-loader/pull/15/commits/41ce0c7e66dd298b1abc4e724d4c6d784ea92373)!

cmoesel commented on test/load.test.ts:
```diff
@@ -55,6 +89,39 @@ const EXT_PKG_RESPONSE = {
+      description: 'None',
+      url: 'https://packages2.fhir.org/packages/hl7.fhir.uv.extensions/1.1.0'
+    },
+    '1.1.2': {
```
Is there a reason you are keeping the `hl7.fhir.uv.extensions` and `hl7.terminology.r4` versions in sync like this? I feel like the unit tests would be more robust if we expect different latest versions from each package instead of the same latest versions from each package. Just in case there was every a bug that mixed packages up.

jafeltra replied:
Oh, good point! No real reason, I just thought it simplified things for my mind reading the tests, but I think that's a good reason to not over simplify it. I'll revert these ones back to a v2 or v3 since that makes sense for the latest version, which is a v4.

jafeltra replied:
Updated in [b53ef0a](https://github.com/standardhealth/fhir-package-loader/pull/15/commits/b53ef0aa0ad1e6b1b6377ce9ce33d6eda38f582d)

cmoesel reviewed (changes requested):
Oops. I guess I had more to say!

jafeltra committed:
Use different versions for packages2 package

jafeltra committed:
Add error for incorrect wildcard uses and throw in mergeDependency and lookUpLatestPatchVersion

cmoesel reviewed (changes requested):
This works quite well!

In testing this, I did notice that running `npm install` reports 3 medium-level vulnerabilities, all of which can be fixed using `npm audit fix`. Since one of them is from the library you introduced (`semver`), I think it makes sense to fix them in this PR.

jafeltra committed:
npm audit fix

jafeltra commented:
[5ba3dfd](https://github.com/standardhealth/fhir-package-loader/pull/15/commits/5ba3dfdd6b3d4f9df2e2f5b46a1e034127b7efc2) - npm audit fixed!

cmoesel reviewed (approved):
APPROVED!
